### PR TITLE
Make TE, TR... to tensors in the header

### PR DESCRIPTION
**Pull request by ckolbPTB**
Not sure about this line here:
https://github.com/PTB-MR/mrpro/blob/9183c2e7573b17106b2f72034bf97429d573ea2f/src/mrpro/data/_IHeader.py#L94

Are we happy with an empty tensor or would it be better to return None?

### CODE REVIEW

File: src/mrpro/data/_KHeader.py [hunk 1]
```diff
@@ -108,11 +108,11 @@ def from_ismrmrd(
...
 
         # Conversion functions for units
-        def sequence_ms_to_s(ms: Sequence[float]) -> Sequence[float]:
-            return [el / 1000 for el in ms]
+        def ms_to_s(ms: torch.Tensor) -> torch.Tensor:
+            return ms / 1000
 
-        def sequence_deg_to_rad(deg: Sequence[float]) -> Sequence[float]:
-            return [el * pi / 180 for el in deg]
+        def deg_to_rad(deg: torch.Tensor) -> torch.Tensor:
```
Comment (fzimmermann89) [inline src/mrpro/data/_KHeader.py]: we can remove this and use torch.deg2rad

Comment (fzimmermann89) [reply src/mrpro/data/_KHeader.py]: (or call deg2rad in this)

**Comment by fzimmermann89** [comment]
> Not sure about this line here:
> 
> https://github.com/PTB-MR/mrpro/blob/9183c2e7573b17106b2f72034bf97429d573ea2f/src/mrpro/data/_IHeader.py#L94
> 
> Are we happy with an empty tensor or would it be better to return None?

Similiar question with KHeader (as the values are not mandatory in ismrmd headers)

Empty list was clear... Empty tensor maybe not?
I would be ok with them being None.

Disadvantage 
  all functions using header.ti etc have to deal with None or mypy will complain.

Advantage
   all functions using header.ti etc have to deal with None or mypy will complain.

**Comment by ckolbPTB** [comment]
> all functions using header.ti etc have to deal with None or mypy will complain.

Wouldn't this mean that we would need something like:

```
class InversionRecovery(SignalModel[torch.Tensor, torch.Tensor]):
    def __init__(self, ti: float | torch.Tensor | None):
         if ti is None:
               raise ValueError
```
This does not look nice so maybe more of a disadvantage rather than advantage

**Comment by fzimmermann89** [comment]
So far we do not test if the signal models work with empty tensors.
T1 would then return an empty tensor as well.
And the fourier operator has to work with an empty tensor etc...

So at some point we might just detect that and throw an exception?
Than mypy could at least warn check if we handle the None Case..

We could also say its neither the job of kdata nor of the signal model to check that ti is not None.
This should be handled by in the code creating the signal operator.
At least if one wants to write correct code that mypy checks.

So for simple reconstruction scripts we don't have to check this at all.
But for a good example or code within mrpro,we have to catch the case that ti is not set before creating the operator.
And mypy would remind us that we need to handle missing ti times somehow. 

So.... thinking about it, my preference would be:
- Set ti to None if missing
- Do not allow None as the ti argument in the signal models

**Comment by fzimmermann89** [comment]
https://mypy-play.net/?mypy=latest&python=3.12&gist=2b31f07eb7db70180370831e13667cc8

but this will not work for "empty tensor"

**Comment by ckolbPTB** [comment]
@fzimmermann89 is this roughly what you had in mind? 

For KHeader I added default values for all fields which are not mandatory in the ismrmrd header. Now this test https://github.com/PTB-MR/mrpro/blob/6338d8127eddb8f7df1ac4aa1aecdbde9d187f78/tests/data/test_kheader.py#L23
passes except for `trajectory`. `trajectory` is defined as `KTrajectoryCalculator` which of course will never be available from ISMRMRD. In our tests we are overwriting trajectory with `xsd.trajectoryType('other')` which seems to be a bug.

Do we really need KTrajectoryCalculator` in the header? We wrote all our code such that we don't need to know the trajectory...

**Comment by ckolbPTB** [comment]
As far as I remember the reason to not use None in KHeader before was, that we said, we want to ensure that certain parameters are defined and we don't have to worry about them later. So we could go ahead with the current approach and allow for None or the other option would be to enforce that all values have to be set in IHeader (using also overwrite and defaults in the same way as for KHeader). Not sure which option I prefer....

**Comment by fzimmermann89** [comment]
> @fzimmermann89 is this roughly what you had in mind?
> 
> For KHeader I added default values for all fields which are not mandatory in the ismrmrd header. Now this test
> 
> https://github.com/PTB-MR/mrpro/blob/6338d8127eddb8f7df1ac4aa1aecdbde9d187f78/tests/data/test_kheader.py#L23
> 
> 
> passes except for `trajectory`. `trajectory` is defined as `KTrajectoryCalculator` which of course will never be available from ISMRMRD. In our tests we are overwriting trajectory with `xsd.trajectoryType('other')` which seems to be a bug.
> Do we really need KTrajectoryCalculator` in the header? We wrote all our code such that we don't need to know the trajectory...

We don't __need__ it, but @schuenke and I thought it might be reallyuseful to know what Trajectory it actually is. 
Yes, the tests is wrong. at some point we should try to enable mypy in the tests as well...


 
 

> As far as I remember the reason to not use None in KHeader before was, that we said, we want to ensure that certain parameters are defined and we don't have to worry about them later. 

so ti would still be None, but we enforce to have a dicom with te, tr, flipangle set (or using an overwrite) ?

None is better than an empty tensor.
But I have no experience how often te,tr,,fa values are missing in dicoms.. Will this be annoying as we always have to guess them and potentially set them to the wrong values?

**Comment by ckolbPTB** [comment]
> None is better than an empty tensor.

Agreed! and I would also say that None is better than a wrong value. 

> But I have no experience how often te,tr,,fa values are missing in dicoms

TE and FA are probably always set, TR maybe not for single-shot stuff like EPI. Main problem is TI. I would prefer to default TI to None if it is not defined rather than forcing the user to set it by hand to a (potentially) wrong value.

**Comment by fzimmermann89** [comment]
So, do we enforce FA and TE.
And default to None for TR and T1?

**Comment by ckolbPTB** [comment]
Maybe better to default to None for TE, TR, FA and TI. I don't see much benefit to treat TE and FA differently to TR and TI but it might lead to unforeseen restrictions (e.g. a spiral does not technically have a TE time, we usually "misuse" TE for the time between RF-pulse and start of the acquisition).

My suggestion would be to default to None for all of them. This would mean that apart from `KTrajectoryCalculator` we can create a KHeader from a minimal ISMRMRD file.

File: src/mrpro/data/_KHeader.py [hunk 2]
```diff
@@ -56,15 +56,15 @@ class KHeader(MoveDataMixin):
...
-    n_coils: int
     acq_info: AcqInfo
-    datetime: datetime.datetime
     h1_freq: float
-    te: list[float]
-    ti: list[float]
-    fa: list[float]
-    tr: list[float]
-    echo_spacing: list[float]
+    n_coils: int | None = None
```
Comment (schuenke) [inline src/mrpro/data/_KHeader.py]: we don't test for n_coils = None yet, do we?
Could imagine that this leads to problems...

**Review by schuenke — APPROVED** [review]

File: src/mrpro/data/_KHeader.py (hunk 2 above)
Comment (ckolbPTB) [reply src/mrpro/data/_KHeader.py]: Probably, but I don't think we are using this parameter anywhere at the moment.

File: src/mrpro/data/_KHeader.py [hunk 3]
```diff
@@ -79,9 +79,11 @@ class KHeader(MoveDataMixin):
     trajectory_description: TrajectoryDescription = dataclasses.field(default_factory=TrajectoryDescription)
 
     @property
-    def fa_degree(self) -> list[float]:
+    def fa_degree(self) -> torch.Tensor | None:
         """Flip angle in degree."""
-        return [el / pi * 180 for el in self.fa]
+        if not self.fa:
```
Comment (fzimmermann89) [inline src/mrpro/data/_KHeader.py]: is it on purpose that this will return None if the flip_angle is set to zero?

Comment (fzimmermann89) [reply src/mrpro/data/_KHeader.py]: also, we have now the same if check twice.
can we make it 
```
if self.fa is None:
   warn
   return None
else:
    return torch.rad2deg(self.fa)
```

**Review by fzimmermann89 — DISMISSED** [review]
see comments

**Review by fzimmermann89 — COMMENTED** [review]
see comments

**Pull request merged**